lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1461 @@
+//===- InferResets.cpp - Infer resets and add async reset -------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InferResets pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/FieldRef.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/EquivalenceClasses.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/TinyPtrVector.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "infer-resets"
+
+using llvm::BumpPtrAllocator;
+using llvm::MapVector;
+using llvm::SmallDenseSet;
+using llvm::SmallSetVector;
+using mlir::FailureOr;
+using mlir::InferTypeOpInterface;
+using mlir::WalkOrder;
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Traits
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+template <>
+struct DenseMapInfo<FModuleOp> {
+  static inline FModuleOp getEmptyKey() {
+    return FModuleOp::getFromOpaquePointer(DenseMapInfo<void *>::getEmptyKey());
+  }
+  static inline FModuleOp getTombstoneKey() {
+    return FModuleOp::getFromOpaquePointer(
+        DenseMapInfo<void *>::getTombstoneKey());
```

Comment:
Nit, but I'd prefer these to be implemented in terms of `DenseMapInfo<Operation *>::` instead of the void* one for consistency.

Reply:
:+1: